controller/src/beerocks/master/son_master_thread.cpp

```diff
@@ -1679,6 +1680,53 @@ bool master_thread::handle_cmdu_1905_topology_response(const std::string &src_ma
         }
     }
 
+    // The reported neighbors list might not be correct since the reporting al_mac hasn't received 
+    // a Topology Discovery from its neighbors yet. Therefore, remove a neighbor node only if more
+    // than 65 seconds have passed since we added this node. This promise that the reported al_mac
+    // will get the Topology Discovery messages from its neighbors and add them to the report.
+    auto last_state_change_timestamp = database.get_last_state_change(src_mac);
+    if (last_state_change_timestamp + std::chrono::seconds(65) < std::chrono::steady_clock::now()) {
+        LOG(TRACE) << "Checking if one of " << src_mac << " is no longer connected";
+        std::unordered_set<sMacAddr> reported_neighbor_al_macs;
+        auto tlv1905NeighborDeviceList = cmdu_rx.getClassList<ieee1905_1::tlv1905NeighborDevice>();
+        for (const auto &tlv1905NeighborDevice : tlv1905NeighborDeviceList) {
+            if (!tlv1905NeighborDevice) {
+                LOG(ERROR) << "ieee1905_1::tlv1905NeighborDevice has invalid pointer";
+                return false;
+            }
+
+            const auto neighbor_al_mac_tuple = tlv1905NeighborDevice->mac_al_1905_device(0);
+            if (!std::get<0>(neighbor_al_mac_tuple)) {
+                LOG(ERROR) << "Getting al_mac element has failed";
+            }
+
+            auto &neighbor_al_mac = std::get<1>(neighbor_al_mac_tuple).mac;
+            LOG(DEBUG) << "Inserting reported neighnor " << neighbor_al_mac << " to the list";
+            reported_neighbor_al_macs.insert(neighbor_al_mac);
+        }
+
+        auto neighbor_al_macs_on_db = database.get_1905_1_neighbors(al_mac);
+        LOG(TRACE) << "Comparing reported neighbors to neighbors on the database, neighbors_on_db="
+                   << neighbor_al_macs_on_db.size();
+        for (const auto &neighbor_al_mac_on_db : neighbor_al_macs_on_db) {
+            // If reported al_mac is on the db skip it, otherwise remove the node.
+            LOG(DEBUG) << "Checks if al_mac " << al_mac << " neighbor " << neighbor_al_mac_on_db
+                       << " is reported in this message";
+            if (reported_neighbor_al_macs.find(neighbor_al_mac_on_db) !=
+                reported_neighbor_al_macs.end()) {
+                continue;
+            }
+            LOG(DEBUG) << "known neighbor al_mac  " << neighbor_al_mac_on_db
+                       << " is not reported on 1905 Neighbor Device TLV, removing the al_mac node";
+
+            std::string neighbor_al_mac_on_db_str =
+                network_utils::mac_to_string(neighbor_al_mac_on_db);
+            auto backhhaul_mac = database.get_node_parent(neighbor_al_mac_on_db_str);
+            son_actions::handle_dead_node(backhhaul_mac, database.get_node_parent(backhhaul_mac),
+                                          database, cmdu_tx, tasks);
+        }
+    }
```

Comment:
I think we should also send a topology query to all the reported neighbours that are not currently in the database, and add them (actually I'd add them on reception of the topology response, not right away).

Also, I'm missing actually adding the neighbour information in the database. I think now this is only done with the intel slave join, and in the non-intel slave join we assume it's connected directly to the gateway. Now would be a good time to fix that (or add an issue to fix it for M3).

Reply:
Currently, we send a query to Agents when receiving M1 from them.
It might be a good idea to send a query on new neighbor that has been discovered, and send the M1 after that instead. Need to think about it...
Anyway, what you are talking about is not in the task scope.